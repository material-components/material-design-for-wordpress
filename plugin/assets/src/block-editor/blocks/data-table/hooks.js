/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { createHigherOrderComponent } from '@wordpress/compose';
import { select, dispatch } from '@wordpress/data';
import domReady from '@wordpress/dom-ready';
import { addFilter } from '@wordpress/hooks';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import DataTableSave from './save';
import DataTableEdit from './edit';

export const isMaterialTableBlock = ( name, attributes ) => {
	return (
		'core/table' === name &&
		attributes &&
		attributes.className &&
		-1 !== attributes.className.indexOf( 'material' )
	);
};

/**
 * Maybe use material data table edit component.
 *
 * @param {WPElement} element    Block save result.
 * @param {WPBlock}   blockType  Block type definition.
 * @param {Object}    attributes Block attributes.
 */
export const withDataTableEdit = createHigherOrderComponent( BlockEdit => {
	return props => {
		if ( isMaterialTableBlock( props.name, props.attributes ) ) {
			return (
				<>
					<DataTableEdit { ...props } />
				</>
			);
		}

		return <BlockEdit { ...props } />;
	};
}, 'withMaterialDataTableEdit' );

/**
 * Maybe use material data table save component.
 *
 * @param {WPElement} element    Block save result.
 * @param {WPBlock}   blockType  Block type definition.
 * @param {Object}    attributes Block attributes.
 */
export const save = ( element, blockType, attributes ) => {
	if ( isMaterialTableBlock( blockType.name, attributes ) ) {
		return <DataTableSave { ...{ attributes } } />;
	}

	return element;
};

/**
 * Add `Material` style.
 *
 * @param {Object} settings Block settings.
 * @param {string} name     Block name.
 */
export const addMaterialStyle = ( settings, name ) => {
	if ( 'core/table' === name ) {
		settings.styles = [
			{
				name: 'material',
				label: __( 'Material', 'material-design' ),
				isDefault: true,
			},
			{
				name: 'regular',
				label: __( 'Regular', 'material-design' ),
			},
			settings.styles[ 1 ],
		];
	}

	return settings;
};

addFilter(
	'blocks.registerBlockType',
	'material/data-table-style',
	addMaterialStyle
);

addFilter(
	'editor.BlockEdit',
	'material/data-table-edit',
	withDataTableEdit,
	1
);

addFilter( 'blocks.getSaveElement', 'material/data-table-save', save );

domReady( () => {
	const stylePreferences = select( 'core/edit-post' ).getPreference(
		'preferredStyleVariations'
	);

	if (
		stylePreferences &&
		( ! stylePreferences[ 'core/table' ] ||
			'material' !== stylePreferences[ 'core/table' ] )
	) {
		dispatch( 'core/edit-post' ).updatePreferredStyleVariations(
			'core/table',
			'material'
		);
	}
} );
