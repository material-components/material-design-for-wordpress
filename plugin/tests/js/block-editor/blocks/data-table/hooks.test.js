/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';

/**
 * WordPress dependencies
 */
import { addFilter } from '@wordpress/hooks';

/**
 * Internal dependencies
 */
import {
	isMaterialTableBlock,
	withDataTableEdit,
	addMaterialStyle,
	save,
} from '../../../../../assets/src/block-editor/blocks/data-table/hooks';

jest.mock(
	'../../../../../assets/src/block-editor/blocks/data-table/edit',
	() => ( {
		__esModule: true,
		default: () => <div>Material Data Table Edit Component</div>,
	} )
);

// Mock wp.hooks.addFilter.
jest.mock( '@wordpress/hooks', () => {
	return {
		addFilter: jest.fn(),
	};
} );

// Mock wp.blocks.registerBlockStyle.
jest.mock( '@wordpress/blocks', () => {
	return {
		registerBlockStyle: jest.fn(),
	};
} );

// Mock wp.data.
jest.mock( '@wordpress/block-editor', () => ( {
	RichText: jest.fn(),
	getColorClassName: jest.fn(),
	createCustomColorsHOC: Comp => () => () => Comp,
} ) );

// Mock wp.data.
jest.mock( '@wordpress/data', () => ( {
	combineReducers: jest.fn(),
	registerStore: jest.fn(),
	select: store => {
		switch ( store ) {
			case 'core/edit-post':
				return {
					getPreference: () => {
						return {
							'core/table': 'regular',
						};
					},
				};

			default:
				return {};
		}
	},
	withSelect: () => () => {
		return {};
	},
	dispatch: store => {
		switch ( store ) {
			case 'core/edit-post':
				return {
					updatePreferredStyleVariations: jest.fn(),
				};

			default:
				return {};
		}
	},
} ) );

const attributes = {
	hasFixedLayout: true,
	head: [
		{
			cells: [
				{
					content: 'Version',
					tag: 'th',
				},
				{
					content: 'Jazz Musician',
					tag: 'th',
				},
				{
					content: 'Release Date',
					tag: 'th',
				},
			],
		},
	],
	body: [
		{
			cells: [
				{
					content: '5.2',
					tag: 'td',
				},
				{
					content: 'Jaco Pastorius',
					tag: 'td',
				},
				{
					content: 'May 7, 2019',
					tag: 'td',
				},
			],
		},
		{
			cells: [
				{
					content: '5.1',
					tag: 'td',
				},
				{
					content: 'Betty Carter',
					tag: 'td',
				},
				{
					content: 'February 21, 2019',
					tag: 'td',
				},
			],
		},
		{
			cells: [
				{
					content: '5.0',
					tag: 'td',
				},
				{
					content: 'Bebo ValdÃ©s',
					tag: 'td',
				},
				{
					content: 'December 6, 2018',
					tag: 'td',
				},
			],
		},
	],
	foot: [
		{
			cells: [
				{
					content: 'Footer Col 1',
					tag: 'td',
				},
				{
					content: 'Footer Col 2',
					tag: 'td',
				},
				{
					content: 'Footer Col 3',
					tag: 'td',
				},
			],
		},
	],
	caption: 'Material Data Table',
	className: 'is-style-material',
};

describe( 'Data Table Filters', () => {
	it( 'filters should be added', () => {
		expect( addFilter ).toHaveBeenCalledTimes( 3 );

		// eslint-disable-next-line jest/prefer-strict-equal
		expect( addFilter.mock.calls[ 0 ] ).toEqual( [
			'blocks.registerBlockType',
			'material/data-table-style',
			addMaterialStyle,
		] );

		// eslint-disable-next-line jest/prefer-strict-equal
		expect( addFilter.mock.calls[ 1 ] ).toEqual( [
			'editor.BlockEdit',
			'material/data-table-edit',
			withDataTableEdit,
			1,
		] );

		// eslint-disable-next-line jest/prefer-strict-equal
		expect( addFilter.mock.calls[ 2 ] ).toEqual( [
			'blocks.getSaveElement',
			'material/data-table-save',
			save,
		] );
	} );

	it( 'addMaterialStyle should add material style', () => {
		let settings = addMaterialStyle( { styles: [] }, 'core/paragraph' );
		expect( settings.styles ).toHaveLength( 0 );

		settings = addMaterialStyle(
			{
				styles: [
					{
						name: 'regular',
						label: 'Default',
						isDefault: true,
					},
					{ name: 'stripes', label: 'Stripes' },
				],
			},
			'core/table'
		);
		expect( settings.styles ).toHaveLength( 3 );
	} );

	it( 'isMaterialTableBlock should return correct value', () => {
		expect(
			isMaterialTableBlock( 'core/table', { className: 'is-style-material' } )
		).toStrictEqual( true );

		expect(
			isMaterialTableBlock( 'core/table', { className: 'is-style-regular' } )
		).toStrictEqual( false );

		expect(
			isMaterialTableBlock( 'core/paragraph', {
				className: 'is-style-material',
			} )
		).toStrictEqual( false );
	} );

	it( 'withDataTableEdit should include DataTableEdit', () => {
		const Component = withDataTableEdit( () => {
			return <div>Original Edit component</div>;
		} );
		let wrapper = render(
			<Component { ...{ name: 'core/table', attributes } } />
		);

		expect( wrapper ).toMatchSnapshot();

		wrapper = render(
			<Component { ...{ name: 'core/paragraph', attributes } } />
		);

		expect( wrapper ).toMatchSnapshot();
	} );

	it( 'save should invoke DataTableSave', () => {
		const Component = () => {
			return <div>Original Save component</div>;
		};
		const element = <Component />;
		let saveElement = save( element, { name: 'core/table' }, attributes );

		expect( saveElement ).toMatchSnapshot();

		saveElement = save( element, { name: 'core/paragraph' }, attributes );

		expect( saveElement ).toMatchSnapshot();
	} );
} );
