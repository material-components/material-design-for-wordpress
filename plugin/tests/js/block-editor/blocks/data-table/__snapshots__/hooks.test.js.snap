// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Table Filters save should invoke DataTableSave 1`] = `
<Save
  attributes={
    Object {
      "body": Array [
        Object {
          "cells": Array [
            Object {
              "content": "5.2",
              "tag": "td",
            },
            Object {
              "content": "Jaco Pastorius",
              "tag": "td",
            },
            Object {
              "content": "May 7, 2019",
              "tag": "td",
            },
          ],
        },
        Object {
          "cells": Array [
            Object {
              "content": "5.1",
              "tag": "td",
            },
            Object {
              "content": "Betty Carter",
              "tag": "td",
            },
            Object {
              "content": "February 21, 2019",
              "tag": "td",
            },
          ],
        },
        Object {
          "cells": Array [
            Object {
              "content": "5.0",
              "tag": "td",
            },
            Object {
              "content": "Bebo ValdÃ©s",
              "tag": "td",
            },
            Object {
              "content": "December 6, 2018",
              "tag": "td",
            },
          ],
        },
      ],
      "caption": "Material Data Table",
      "className": "is-style-material",
      "foot": Array [
        Object {
          "cells": Array [
            Object {
              "content": "Footer Col 1",
              "tag": "td",
            },
            Object {
              "content": "Footer Col 2",
              "tag": "td",
            },
            Object {
              "content": "Footer Col 3",
              "tag": "td",
            },
          ],
        },
      ],
      "hasFixedLayout": true,
      "head": Array [
        Object {
          "cells": Array [
            Object {
              "content": "Version",
              "tag": "th",
            },
            Object {
              "content": "Jazz Musician",
              "tag": "th",
            },
            Object {
              "content": "Release Date",
              "tag": "th",
            },
          ],
        },
      ],
    }
  }
/>
`;

exports[`Data Table Filters save should invoke DataTableSave 2`] = `<Component />`;

exports[`Data Table Filters withDataTableEdit should include DataTableEdit 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        Material Data Table Edit Component
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      Material Data Table Edit Component
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Data Table Filters withDataTableEdit should include DataTableEdit 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        Material Data Table Edit Component
      </div>
    </div>
    <div>
      <div>
        Original Edit component
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      Original Edit component
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
